### PR TITLE
[SPARK-47927][SQL][FOLLOWUP] fix ScalaUDF output nullability

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/46156 , to fix the wrong nullability of ScalaUDF output.

### Why are the changes needed?

fix nullability

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cloud-fan** [comment]
cc @eejbyfeldt  @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -344,10 +344,11 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       new ResolveHints.RemoveAllHints),
     Batch("Nondeterministic", Once,
       PullOutNondeterministic),
+    Batch("ScalaUDF", Once,
+      HandleNullInputsForUDF),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:348]: This rule adds null handling for ScalaUDF, so even if the `ScalaUDF`'s `nullable` is false, the final expression (if-else) can also return null, so we need to run `UpdateAttributeNullability` after it.

**Review by yaooqinn — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:348]: But we also need to run it before? Or how could we correctly handle null inputs to udf if we do not have the correct nullability for the inputs?

**Comment by cloud-fan** [comment]
@eejbyfeldt I did consider this approach before. However, the null check happens outside of `ScalaUDF`. If we do invoke the UDF, `ScalaUDF.nullable = false` means the UDF result won't be null.

**Comment by eejbyfeldt** [comment]
> @eejbyfeldt I did consider this approach before. However, the null check happens outside of `ScalaUDF`. If we do invoke the UDF, `ScalaUDF.nullable = false` means the UDF result won't be null.

Could you expand on how to trigger this error? I am not sure I am following. (But I do agree that it feels a bit off the the implementation of `nullable` does not match implementation of the execution)


Here is a code that still produces incorrect result with this branch (and master):
```
scala> val f = udf[Int, Int](_ + 1)
scala> sc.parallelize(Seq(Some(1), None)).toDS.select(f($"value").as("t"), f($"t")).show()
+----+-----------------------------+
|   t|UDF(lateralAliasReference(t))|
+----+-----------------------------+
|   2|                            3|
|NULL|                            1|
+----+-----------------------------+
```

To solve (while keeping the `ScalaUDF.nullable` implementation) this we would probably need to make the update nullability batch be something like
```
    Batch("UpdateNullability", fixedPoint,
      HandleNullInputsForUDF,
      UpdateAttributeNullability),
```

**Comment by cloud-fan** [comment]
@eejbyfeldt good catch! I've fixed it and added a test

**Review by eejbyfeldt — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the reviews, merging to master/3.5!

**Pull request closed without merging**